### ISSUE TITLE
Axios hangs when POST an empty buffer

### ISSUE THREAD

**Issue by amitzur** [axios/axios#1701]
#### Summary

When sending a 'POST' or 'PUT' request with an empty buffer, the promise doesn't resolve or reject and the server doesn't get the request.
The following example demonstrates the issue:
```
const axios = require('axios');
const http = require('http');

const server = http.createServer((req, res) => {
  console.log('received request');
  res.writeHead(200, {'Content-Type': 'text/plain'});
  res.end('okay');
});

['listening', 'error', 'connection', 'close'].forEach(ev => {
  server.on(ev, function() {
    console.log(ev);
  });
});

server.listen(7373, async () => {
  await axios({
    url: `http://localhost:7373/hb`,
    method: 'post',
    data: Buffer.from(''),
  });

  await server.close();
});
```
The output for the above is:
```
listening
connection
```
and the process hangs. If you replace `Buffer.from('')` with just `''` then it works properly and the output is:
```
listening
connection
request received
close
```

#### Context

- axios version: 0.18.0
- Environment: node v10.6.0

**Comment by delirius325** [comment]
Hi there @amitzur,

From what I gather (and from the Node.js documentation), ```Buffer.from('')``` is not returning a string. Therefore you should write ```Buffer.from('').toString()```and everything should work as expected 😄 !

[See Node.JS documentation](https://nodejs.org/api/buffer.html#buffer_class_method_buffer_from_string_encoding)

Cheers,

**Comment by amitzur** [comment]
Hey @delirius325 ,

AKAICT node buffers should be supported as well. From the readme:
```
  // `data` is the data to be sent as the request body
  // Only applicable for request methods 'PUT', 'POST', and 'PATCH'
  // When no `transformRequest` is set, must be of one of the following types:
  // - string, plain object, ArrayBuffer, ArrayBufferView, URLSearchParams
  // - Browser only: FormData, File, Blob
  // - Node only: Stream, Buffer
```

For example this line of code specifically handles buffers:
https://github.com/axios/axios/blob/master/lib/adapters/http.js#L41

**Comment by delirius325** [comment]
You're right, it should technically support it without having to parse the Buffer as a string.

```javascript
if (data && !utils.isStream(data)) {
      if (Buffer.isBuffer(data)) {
        // Nothing to do...
      } else if (utils.isArrayBuffer(data)) {
        data = Buffer.from(new Uint8Array(data));
      } else if (utils.isString(data)) {
        data = Buffer.from(data, 'utf-8');
      } else {
        return reject(createError(
          'Data after transformation must be a string, an ArrayBuffer, a Buffer, or a Stream',
          config
        ));
      }

      // Add Content-Length header if data exists
      headers['Content-Length'] = data.length;
    }
```
https://github.com/axios/axios/blob/master/lib/adapters/http.js#L40

I will try to debug it on my local machine, I'll push a PR if I find out what's going on 😄 !
Cheers,

**Comment by amitzur** [comment]
thanks!

**Comment by delirius325** [comment]
I created a PR (#1703) for this request. Basically converting the empty Buffer into a string.

Cheers!

**Comment by amitzur** [comment]
Thanks for the quick turnaround!
Do you have an idea why this would happen though?

**Comment by delirius325** [comment]
I'm not done investigating, although the problem seems to be lying in the ```follow-redirects``` module that axios uses for its requests. Can't seem to find the issue yet though, that's why my PR is just a work-around for now 😄 !

**Comment by delirius325** [comment]
@amitzur The problem was indeed related to the ```follow-redirects```module used by axios. I made a PR to fix the problem in that library (which is the source of the issue) instead of patching the problem into axios.

**Comment by delirius325** [comment]
Hi @amitzur ,

Issue has been fixed in ```follow-redirects```. The problem was deeper than just the module itself, it goes to Node's http module which does not trigger the callback on empty strings/buffers.

https://github.com/nodejs/node/blob/v10.7.0/lib/_http_outgoing.js#L608
https://nodejs.org/api/stream.html#stream_writable_write_chunk_encoding_callback

Since the behaviour is weird, it's been reported here (following this investigation)  : https://github.com/nodejs/node/issues/22066 

This issue can be closed as it is not related directly to```axios``` and it's been fixed at the source.
Thanks a lot 😄,

**Comment by delirius325** [comment]
Ruben, the person who opened the issue is also the biggest contributor of ```follow-redirects```. So when I made my PR to the module, he was curious if it was solely related to follow-redirects or directly to Node. When he discovered that it was directly related to Node he decided to open up an issue 😃 .

**Comment by amitzur** [comment]
Fantastic work, thank you so much :)
